### DISCUSSION TITLE
General directory-based partitioning

### DISCUSSION THREAD

**Discussion by plaflamme** [duckdb/duckdb#6897] [Ideas]
## Current State

Currently, DuckDB supports Hive-style partitioning of datasets. That is, when the dataset's folders structure conforms to Hive's conventions (i.e.: `/path/to/dataset/partition_a=123/partition_b=abc/0001.parquet`), then specifying `hive_partitioning=1` will allow DuckDB to leverage this and use predicate pushdown and such.

Encoding partitions in path components is a common way to split a dataset, but conforming to Hive's conventions is not strictly required to achieve the same benefits. For example, a daily dataset could use `/path/2023/02/01.parquet` or `/path/2023-02-01.parquet` to achieve the same result.

Support for non-Hive-conforming paths is the proposition laid out here.

## Proposition

Allow specifying the partition column's name(s) by using path "templates" to locate files.

The column name could be enclosed in curlies or similar and. For example: `/path/{year}/{month}/{day}.parquet` would match `/path/2023/02/01.parquet` and assign `2023` to a `year` column, `02` to a `month` column and `01` to a `day` column.

To keep things simple, all columns would be parsed as `VARCHAR` similar to how Hive partitioning works.

The path template could also be specified separately from the globbing pattern, e.g.:

```
read_parquet('/path/**', partitioning='/path/{year}/{month}/{day}')
```

This could also be extended to support specifying the columns datatype (as long as it can be cast from STRING?), e.g.: `/path/{year:INT}/{month:INT}` or `/path/{dt:DATE}.parquet`

## Alternative

Arrow "datasets" already support this. For example, see [here](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.partitioning.html#pyarrow-dataset-partitioning) for the Python API into this feature.

So an alterantive is to support something similar to Arrow's `DirectoryPartitioning`. It's not as general, but probably close enough. Perhaps something like (not sure about the schema syntax):

```
read_parquet('/path/**', directory_partitioning=(dt DATE, value INT))
```

## Alternative 2

Supporting [Arrow `Dataset`](https://arrow.apache.org/docs/python/dataset.html) themselves could be considered as well. This is already possible from, say python, but supporting this in SQL would be nice for cases like `dbt` where you need to build a view from SQL, e.g.:

```
CREATE VIEW my_dataset AS SELECT * FROM arrow_dataset('/path/to/dataset', directory_partittioning=...)
```

**Comment by plaflamme** [comment]
I've been putting more thoughts into this and I've been realizing that what I really want is something akin to `CREATE EXTERNAL TABLE` in other warehousing systems.

The syntax differs from system to system, but they mostly all offer the same kinds of features:

* multiple file format (csv, parquet, etc.)
* partitioning (e.g.: `PARTITION BY col1,col2`)
* multiple locations / filesystems (local, s3, gcs, etc.)
* custom table properties (arbitrary key/value pairs e.g.: for specifying compression options when writing parquet files)

A lot of these are already supported by `read_parquet`, `read_csv` and such, but there are a few limitations:

* they only allow reading from those files
* they are not DDL statements which means they don't "contribute" the the schema, one must combine this with `CREATE VIEW` or similar

Having a formal concept of an "external table" would allow the following things:

* writing to the table could use plain DML statements like `INSERT` (though supporting `DELETE` and `UPDATE` might be challenging)
* more granular control over the configuration of the table
* extensible without "syntax changes" (supporting a new file format would not require a change to the DDL syntax)

My particular use case is that I want to use `dbt` to build my `duckdb` warehouse. I want to and add some tables that are backed by arrow datasets, but in a way that DuckDB doesn't allow me to today. For example, using "directory" partitioning as mentioned in "Alternative" above. I can do that using the Python API, by doing something like

```python
dataset = ds.dataset(...)
con = duckdb.connect()
con.register("my_dataset, dataset)
```

But this does not "persist" between duckdb sessions. So, say I use `dbt` to create a `.duckdb` file with a bunch of views and derived tables and such, those "registered" datasets are no longer available when I re-open my duckdb database since they're not part of the schema. I have to re-register those custom datasets.

Having formal support for external tables (even in read-only mode) would allow me to use something like https://github.com/dbt-labs/dbt-external-tables/ to create **actual** tables in duckdb that would persist when opening the warehouse in a non-dbt context.

We can **almost** do all this now using `dbt-duckdb` except that the `read_parquet(...)` method is too limiting and does not allow controlling the finer details how to "connect" to those files.

I believe formal support for external tables would provide a more consistent way of dealing with these kinds of use-cases.

**Comment by sveniu** [comment]
This would work great for Amazon CloudFront access logs written to S3. An example filename is `/some/prefix/EKBZNHIRCCXUC.2023-12-11-05.ac482179.gz` – so essentially `<static-prefix>.YYYY-mm-dd-HH.<random-hex>.gz`.

Side note: Amazon's own Athena service supports partition projection but only with forward slash as the delimiter. The official recommendation is to rename the CloudFront log files to either make them match Hive-style partitioning, or the projection-compatible slash-only partitioning. Having something like duckdb just work without having to jump through hoops would be much better.

**Comment by MatinF** [comment]
We would be very interested in support for this as well!

**Comment by akdor1154** [comment]
This would open up wild use-cases of duckdb for us - this specifically being a combination of `EXTERNAL TABLE` and `PARTITIONED`. Nothing beyond what's doable now in Postgres with partitioned tables and parquet_fdw, but combine that with duckdb's unique runtime versatility and you have some tantalising possibilities.

**Comment by catkins** [comment]
I'm a big +1 on something in this space, both being able to specify partition path template at query time as well as a `CREATE EXTERNAL TABLE` kind of approach as you can do in hive+trino or athena+glue!

**Comment by borkur** [comment]
Wonder if there's a way to Upvote this great feature improvement :)

**Comment by mootari** [comment]
Support for `PARTITION_PATH` was added to DuckDB 2.0 a few days ago via #25846.

Edit: My bad, that's only for `COPY TO`, not for reading it seems.
